File: go/tut/string.go
package main

import (
	"fmt"
	"strings"
)


func main() {

	// contains
	fmt.Println(strings.Contains("seafood", "foo"))

	// upper case
	fmt.Println(strings.ToUpper("SeaFood"))

	// lower case
	fmt.Println(strings.ToLower("SeaFood"))

	// capitalize (or title)
	fmt.Println(strings.Title("aBc dEf"))

	// trim
	fmt.Println(strings.TrimSpace("  trim  "))

	// split
	fmt.Println(strings.Split("a,b,c,", ","))

	// split lines
	fmt.Println(strings.Split(`line 1
line 2
line 3`, "\n"))

	// is empty
	// - https://stackoverflow.com/a/18595217
	fmt.Println(len("") > 0)
	fmt.Println("" != "")

	// substring
	fmt.Println("seafood"[2:5])

	// reverse
	// - https://stackoverflow.com/a/10030772
	fmt.Println(Reverse("seafood"))

	// length
	fmt.Println(len("seafood"))

	// join
	s := []string{"a", "b", "c"}
	fmt.Println(strings.Join(s, ", "))
	fmt.Println(strings.Join([]string{"item1", "item2", "item3"}, ", "))
}

func Reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes) - 1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

File: javascript/string.js

// contains
console.log('hello'.includes('ll'));
console.log('hello'.includes('LL'));

// upper
console.log('hello'.toUpperCase());

// lower
console.log('Hello'.toLowerCase());

// capital
// https://stackoverflow.com/a/3291856
function capitalize(text) {
    return text.charAt(0).toUpperCase() + text.slice(1).toLowerCase();
}
console.log(capitalize('Hello wOrld'));

// trim
console.log("'" + "    hello      ".trim() + "'");

// split
console.log(" hello world a b c ".split(" "));
console.log(" hello world a b c ".split(" ", 2));

// split lines
// https://stackoverflow.com/a/5035058
var text = "hello\nworld\n\nbye\n";
var lines = text.match(/[^\r\n]+/g);
console.log(lines.length);
for (var i = 0; i < lines.length; i++) {
    var line = lines[i];
    console.log("[" + i + "] " + line);
}

// is empty
var str = "x";
console.log("'" + str + "' is empty? " + (str === ""))
str = "";
console.log("'" + str + "' is empty? " + (str === ""))
str = null;
console.log("'" + str + "' is empty? " + (str === ""))

// substring
console.log("hello world".substring(2, 5));

// reverse
console.log("hello".split("").reverse().join(""));

// length
console.log("hello world".length);

// unicode
console.log("\u2661");
